drivers/disk/flashdisk.c

```diff
@@ -47,45 +47,87 @@ static int disk_flash_access_status(struct disk_info *disk)
 	return DISK_STATUS_OK;
 }
 
-static int disk_flash_access_init(struct disk_info *disk)
+static int flashdisk_init_runtime(struct flashdisk_data *ctx,
+				  const struct flash_area *fap)
 {
-	struct flashdisk_data *ctx;
-	const struct flash_area *fap;
 	int rc;
 	struct flash_pages_info page;
+	off_t offset;
 
-	ctx = CONTAINER_OF(disk, struct flashdisk_data, info);
-
-	rc = flash_area_open(ctx->area_id, &fap);
-	if (rc < 0) {
-		LOG_ERR("Flash area %u open error %d", ctx->area_id, rc);
-		return rc;
-	}
-
-	disk->dev = flash_area_get_device(fap);
-
-	rc = flash_get_page_info_by_offs(disk->dev, ctx->offset, &page);
+	rc = flash_get_page_info_by_offs(ctx->info.dev, ctx->offset, &page);
 	if (rc < 0) {
 		LOG_ERR("Error %d while getting page info", rc);
-		flash_area_close(fap);
 		return rc;
 	}
 
 	ctx->page_size = page.size;
-	LOG_INF("Initialize device %s", disk->name);
+	LOG_INF("Initialize device %s", ctx->info.name);
 	LOG_INF("offset %lx, sector size %zu, page size %zu, volume size %zu",
 		(long)ctx->offset, ctx->sector_size, ctx->page_size, ctx->size);
 
+	if (ctx->cache_size == 0) {
+		/* Read-only flashdisk, no flash partition constraints */
+		LOG_INF("%s is read-only", ctx->info.name);
+		return 0;
+	}
+
+	if (ctx->offset != page.start_offset) {
+		LOG_ERR("Partition does not start at page boundary");
+		return -EINVAL;
+	}
+
+	offset = ctx->offset + page.size;
+	while (offset < ctx->offset + ctx->size) {
+		rc = flash_get_page_info_by_offs(ctx->info.dev, offset, &page);
+		if (rc < 0) {
+			LOG_ERR("Error %d while getting page info", rc);
+			return rc;
+		}
+		if (page.size != ctx->page_size) {
+			LOG_ERR("Non-uniform page size is not supported");
+			return rc;
+		}
+		offset += page.size;
+	}
+
+	if (offset != ctx->offset + ctx->size) {
+		LOG_ERR("Last partition page crossess flashdisk boundary");
+		return -EINVAL;
+	}
+
```

Comment:
This part should be made conditional under Kconfig options. In 99% this is wasted clocks, because once the user figures out that only uniform page access can only be used, for writes to work, this is just running for no reason, and ads to boot time.
Most of checks here should be actually conditional, because so far we do not support initializing disks "by hand" from application, which means all the checks are for constants known at build and taken from DTS, and if they are incorrect once they get corrected this code runs for no reason.

Reply:
Added Kconfig option.

Reply:
Unsure if this is correct, what about spi-nor flashes (with spd) with non-uniform block size?

Reply:
@Laczen the user will simply lose data if the Kconfig option is disabled and partition spans across non-uniform blocks. Therefore, I believe the check should be opt-out (i.e. `default y`) instead of opt-in.

Reply:
> Unsure if this is correct, what about spi-nor flashes (with spd) with non-uniform block size?

I do not think there are such, but assuming this is possible, you will detect it during tests.
If you fail to test this, you will have to put extra work to recover your device from field anyway, whether this code runs or not.

Reply:
These devices exist (e.g. s25fs512s with hybrid sector option). All these tests (and also the ones in nvs/littlefs/...) could have been avoided if the partition definition would have a optional `erase-block-size` to overrule whatever is defined in spd or the flash `erase-block-size`. If this was incorrectly set it would have been detected during tests, but doing this was a nogo in the past, remember ? Now Kconfig options are created to skip over some testing that could have been avoided. Nice !

Reply:
> s25fs512s

You mean this: https://www.infineon.com/dgdl/Infineon-[S25FS512S_512_Mb_1.8_V_Serial_Peripheral_Interface_with_Multi-I_O_Non-Volatile_Flash-DataSheet-v14_00-EN.pdf](https://www.infineon.com/dgdl/Infineon-S25FS512S_512_Mb_1.8_V_Serial_Peripheral_Interface_with_Multi-I_O_Non-Volatile_Flash-DataSheet-v14_00-EN.pdf?fileId=8ac78c8c7d0d8da4017d0ed681a356fe)

The device is unusable as block device as it supports at most 8 continuous 4k blocks per each 256kB page, top or bottom.

> Now Kconfig options are created to skip over some testing that could have been avoided. Nice !

Unless the SFD is read at run-time, which requires running code to figure out. Also in case of the device, the hybrid erase thing depends on device type and configuration. If you use the device for file system, without knowing how it works this is design failure.

What is problem with being able to disable code, that takes time of boot sequence and space in flash, that is supposed to be only run before you deploy device, because once you have tested everything this code is pointless.

If a deployed solution is using an incorrect flash configuration, for disk, it will fail to run in field whether this code runs or not.